### ISSUE TITLE
C++ API `torch::nn:functional::_unpool_output_size`: Segfault by unknown memory access.

### ISSUE THREAD

**Issue by Sehun0819** [pytorch/pytorch#106702]
### 🐛 Describe the bug

When the size of `stride` or `padding` is shorter than the size of `kernel_size`, `torch::nn:functional::_unpool_output_size` crashes.

Test code:
```c++
#include <stdint.h>
#include <stddef.h>
#include <c10/util/irange.h>
#include <cassert>
#include <torch/torch.h>

namespace F = torch::nn::functional;
using namespace torch::nn;

int main() {
  try {
    torch::TensorOptions toptions = torch::TensorOptions();

    auto result = F::_unpool_output_size(
      torch::randn({}, toptions),{0},{},{},{});
  } catch (std::exception& e) {
    return -2;
  }

  return 0;
}
```
Error log:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==702702==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x000000406ae7 bp 0x7ffc2be1c650 sp 0x7ffc2be1bfe0 T0)
==702702==The signal is caused by a READ memory access.
==702702==Hint: address points to the zero page.
    #0 0x406ae7 in torch::nn::functional::_unpool_output_size(at::Tensor const&, c10::ArrayRef<long> const&, c10::ArrayRef<long> const&, c10::ArrayRef<long> const&, c10::optional<std::vector<long, std::allocator<long> > > const&) /home/sehoon/pytorch/torch/csrc/api/include/torch/nn/functional/pooling.h:640:35
    #1 0x405a1a in main /home/sehoon/pytorch/test/cpp/reproduce/_unpool_output_size_bo.cpp:14:19
    #2 0x7ff876d66d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #3 0x7ff876d66e3f in __libc_start_main csu/../csu/libc-start.c:392:3
    #4 0x4054a4 in _start (/home/sehoon/pytorch/build/bin/reproduce__unpool_output_size_bo+0x4054a4)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /home/sehoon/pytorch/torch/csrc/api/include/torch/nn/functional/pooling.h:640:35 in torch::nn::functional::_unpool_output_size(at::Tensor const&, c10::ArrayRef<long> const&, c10::ArrayRef<long> const&, c10::ArrayRef<long> const&, c10::optional<std::vector<long, std::allocator<long> > > const&)
==702702==ABORTING
```
[Error location](https://github.com/pytorch/pytorch/blob/1cc002621d23749c6742193ef3c1eda1015480c2/torch/csrc/api/include/torch/nn/functional/pooling.h#L638-L641):
```c++
  for (const auto d : c10::irange(kernel_size.size())) {
    default_size.push_back(
        (input_size[d + 2] - 1) * stride[d] + kernel_size[d] - 2 * padding[d]);
  }
```
Program crashed at `stride[d]` because `stride.size()` is zero. `input_size[d+2]` doesn't look safe as well, but I have no idea how it passed sanitizer.

### Versions

PyTorch version: 2.1.0a0+git416bf4e
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 12.0.1 (git@github.com:starlab-unist/llvm-project.git 82ccd79bfce79353c2bb4c1ab258ebfeb536a67d)
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.9.17 (main, Jul  5 2023, 20:41:20)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-78-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090

Nvidia driver version: 535.86.10
cuDNN version: Probably one of the following:
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn.so.8.9.2
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.2
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.2
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.2
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.2
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.2
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          96
On-line CPU(s) list:             0-95
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
Stepping:                        7
CPU max MHz:                     4000.0000
CPU min MHz:                     1200.0000
BogoMIPS:                        6000.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       1.5 MiB (48 instances)
L1i cache:                       1.5 MiB (48 instances)
L2 cache:                        48 MiB (48 instances)
L3 cache:                        71.5 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-23,48-71
NUMA node1 CPU(s):               24-47,72-95
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.1.0a0+git416bf4e
[conda] mkl                       2023.1.0         h6d00ec8_46342  
[conda] mkl-include               2023.1.0         h06a4308_46342  
[conda] numpy                     1.25.2                   pypi_0    pypi
[conda] torch                     2.1.0a0+git416bf4e           dev_0    <develop>

cc @jbschlosser @albanD @mruberry @walterddr @mikaylagawarecki

**Comment by albanD** [comment]
We should add the appropriate TORCH_CHECK to ensure these size are correct. Feel free to send a PR fixing this!

**Comment by Sehun0819** [comment]
@albanD There is inconsistency between python API and C++ API.

[Python](https://github.com/pytorch/pytorch/blob/ad22f0ffb456fc3f967ad32e09376f7c9cf94a56/torch/nn/functional.py#L896-L897):
```python
for d in range(len(kernel_size)):
    default_size.append((input_size[-len(kernel_size) + d] - 1) * stride[d] + kernel_size[d] - 2 * padding[d])
```
[C++](https://github.com/pytorch/pytorch/blob/1cc002621d23749c6742193ef3c1eda1015480c2/torch/csrc/api/include/torch/nn/functional/pooling.h#L638-L641):
```c++
for (const auto d : c10::irange(kernel_size.size())) {
  default_size.push_back(
      (input_size[d + 2] - 1) * stride[d] + kernel_size[d] - 2 * padding[d]);
}
```
In python frontend, the base index of `input_size` is `len(input_size) - len(kernel_size) - 1` while C++ frontend simply uses `2`. It seems clear than the length of `kernel_size`, `stride`, and `padding` should be same, but I'm not sure about `input_size`.

**Comment by albanD** [comment]
Hey it is very much possible that this code was updated on the python side to support more than just the basic case the C++ code supports.
I also missed that these are private functions. So you should not be calling them directly. Are you able to reproduce this segfault when calling only public APIs?

**Comment by Sehun0819** [comment]
Segfault does not occur on `stride` and `padding` when `_unpool_output_size` is called from `max_unpool*d` API, because lengths of them are automatically set by `ExpandingArray`.

But still,
```c++
for (const auto d : c10::irange(kernel_size.size())) {
  default_size.push_back(
      (input_size[d + 2] - 1) * stride[d] + kernel_size[d] - 2 * padding[d]);
}
```
this C++ code should be fixed as it is not consistent with [document](https://pytorch.org/docs/master/generated/torch.nn.MaxUnpool1d.html?highlight=maxunpool1d#torch.nn.MaxUnpool1d). `input_size[d + 2]` seems like it only considers (N ,C ,*) shapes, while document(and python API) allows (C, *) shapes. I guess replacing `input_size[d + 2]` with `input_size[input_size.size() - kernel_size.size() - 1]` would work.

**Comment by Sehun0819** [comment]
Another thing is, the absence of shape check(of `max_unpool*d`) seems unsafe. Currently any tensor can reach `input_size[d + 2]` and lead to buffer overflow. Existing shape checks([2d](https://github.com/pytorch/pytorch/blob/2764ead4296541b324ae0e8e2b6200eec0e90c4e/aten/src/ATen/native/MaxUnpooling.cpp#L35-L37), [3d](https://github.com/pytorch/pytorch/blob/2764ead4296541b324ae0e8e2b6200eec0e90c4e/aten/src/ATen/native/MaxUnpooling.cpp#L91-L93)) are executed after those things.

**Comment by Sehun0819** [comment]
```c++
#include <stdint.h>
#include <stddef.h>
#include <c10/util/irange.h>
#include <cassert>
#include <torch/torch.h>

namespace F = torch::nn::functional;
using namespace torch::nn;

int main() {
  try {
    torch::TensorOptions toptions = torch::TensorOptions();

    auto foptions =
      F::MaxUnpool1dFuncOptions(torch::ExpandingArray<1>({1}))
        .stride({1})
        .padding({1});

    auto result =
      F::max_unpool1d(
        torch::randn({1}, toptions),
        torch::randn({1}, toptions),
        foptions);
  } catch (std::exception& e) {
    std::cout << "Exception: " << e.what();
  }

  return 0;
}
```
This code executes `input_size[d + 2]` silently where `input_size.size()` is `1` and `d + 2` is `2`. I don't know why ASAN does not catch it, but when I slightly modify [`pooling.h` code](https://github.com/pytorch/pytorch/blob/1cc002621d23749c6742193ef3c1eda1015480c2/torch/csrc/api/include/torch/nn/functional/pooling.h#L638-L641) like below,
```c++
for (const auto d : c10::irange(kernel_size.size())) {
  default_size.push_back(
      //(input_size[d + 2] - 1) * stride[d] + kernel_size[d] - 2 * padding[d]);
      (input_size.at(d + 2) - 1) * stride[d] + kernel_size[d] - 2 * padding[d]);
}
```
`ArrayRef` reveals the invalid array access with below error log.
```
Exception: ArrayRef: invalid index Index = 2; Length = 1
Exception raised from at at /home/sehoon/pytorch-latest-asan/c10/util/ArrayRef.h:226 (most recent call first):
frame #0: <unknown function> + 0x1b0cde (0x7f9361777cde in /home/sehoon/pytorch-latest-asan/build/lib/libc10.so)
frame #1: <unknown function> + 0x1b0c90 (0x7f9361777c90 in /home/sehoon/pytorch-latest-asan/build/lib/libc10.so)
frame #2: <unknown function> + 0x1b0c20 (0x7f9361777c20 in /home/sehoon/pytorch-latest-asan/build/lib/libc10.so)
frame #3: <unknown function> + 0x1b0ad0 (0x7f9361777ad0 in /home/sehoon/pytorch-latest-asan/build/lib/libc10.so)
frame #4: <unknown function> + 0x1b24f6 (0x7f93617794f6 in /home/sehoon/pytorch-latest-asan/build/lib/libc10.so)
frame #5: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x1b7 (0x7f93617757e7 in /home/sehoon/pytorch-latest-asan/build/lib/libc10.so)
frame #6: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x20c (0x7f936176f16c in /home/sehoon/pytorch-latest-asan/build/lib/libc10.so)
frame #7: ./build/bin/reproduce_max_unpool1d() [0x40e20e]
frame #8: ./build/bin/reproduce_max_unpool1d() [0x40c37b]
frame #9: ./build/bin/reproduce_max_unpool1d() [0x40b8f3]
frame #10: ./build/bin/reproduce_max_unpool1d() [0x407413]
frame #11: main + 0xc15 (0x4061b5 in ./build/bin/reproduce_max_unpool1d)
frame #12: <unknown function> + 0x29d90 (0x7f9357c4dd90 in /lib/x86_64-linux-gnu/libc.so.6)
frame #13: __libc_start_main + 0x80 (0x7f9357c4de40 in /lib/x86_64-linux-gnu/libc.so.6)
frame #14: _start + 0x25 (0x4054d5 in ./build/bin/reproduce_max_unpool1d)
```
